### ISSUE TITLE
Arpeggiator? 

### ISSUE THREAD

**Issue by MrHammyMark** [surge-synthesizer/surge#3675]
Great work on the 1.8 update guys!! The filters are sounding much better!!
Any chance of putting an arp into it? 

Thank you

**Comment by mkruselj** [comment]
It's not in our immediate plans, but it's not ruled out completely.

**Comment by MrHammyMark** [comment]
Cool dude - looking forward to seeing what you guys do for 1.9
Maybe an arp in 2.0? =D

Nice one!

**Comment by mkruselj** [comment]
@baconpaul famously says 2.0 will never happen. Something else will happen instead. 🙂

**Comment by MrHammyMark** [comment]
> @baconpaul famously says 2.0 will never happen. Something else will happen instead. 🙂

Surge Xtreme?

**Comment by mkruselj** [comment]
Close. 😀

**Comment by MrHammyMark** [comment]
Such a tease haha

**Comment by XtremeSounds** [comment]
Perhaps if the step modulator could apply quantization (root note and scale locking) it would work well as a arp? and a great modulation source for other things as well.

**Comment by baconpaul** [comment]
It already does! Drag the step with ctrl or alt or shift and it quantizes to steps of 1/12

**Comment by mkruselj** [comment]
Arpeggiator is different from step seq because it reacts differently based on which notes you play in.

**Comment by baconpaul** [comment]
Yeah my only comment is: as a mod source you can already do note quantization with the step sequencer.

I am getting more into the idea of arp and chord memory overlays as midi transforms. But maybe that’s because i dont want to fight emoji rendering right now :)

**Comment by mkruselj** [comment]
Let's not do the arp and chord memory as overlays. We should use the LFO area and down arrow methodology for that. Feels more consistent...

Which is why those two things are in After XT 1.0 milestone 🙂

**Comment by baconpaul** [comment]
And you even add an emoji to the reply. Cruel, man, just cruel.

**Comment by mkruselj** [comment]
Ahahaha xD

**Comment by XtremeSounds** [comment]
Sorry I mean quantize to notes in a scale like C minor etc.

**Comment by baconpaul** [comment]
Yes so for modulation purposes you can do that with the step sequencer now. Set the modulation depth to 12, use ctrl when you drag, lay out c minor. Not an arpegiattor but harmonically accurate modulation with the step sequencer is entirely possible

**Comment by XtremeSounds** [comment]
I think I will just use a third party sequencer.  Thanks

**Comment by baconpaul** [comment]
Yes surge is not a full blown sequencer. Have you checked out stochas though? https://stochas.org/

**Comment by remisharrock** [comment]
> Yes surge is not a full blown sequencer. Have you checked out stochas though? https://stochas.org/

OMG that's the sequencer needed for surge ! So advanced ! Open source ! Probabilistic ! Wonderful :)

**Comment by Lo-lo78** [comment]
Hello,
I'm glad it was already requested!
I wanted to describe a bit what a good arpeggiator should be like.
It would be very nice if there was a global arpeggiator and also for single oscillator.
In this case, if there was an arpeggiator for a single oscillator, you could think of making a pad with two oscillators and a third one that does the arpeggiator.
There would be some very interesting patches.
Greetings!

**Comment by Lo-lo78** [comment]
Obviously, if there are three oscillators arpeggiating with the rate it could be a lot of fun.
Osc 1: rate 1.0
Osc 2: rate 2.0
Osc 3: rate 3.5
And go on like this!

**Comment by mkruselj** [comment]
Arpeggiator would be aMIDI input processor,  not a per oscillator thing. It would be a separate modulator,  not a part of the LFO.

**Comment by Lo-lo78** [comment]
Could you think of doing it for scenes?
Scenes a with the arpeggiator and scenes b without the arpeggiator.

**Comment by Lo-lo78** [comment]
@mkruselj 
I have an idea for the individual oscillators.
For the global one, midi is fine. For single oscillators one might think of using frequency.
Just like 8-bit computers did.
You know the Commodore 64?
The arpeggiators were made following the frequency of the oscillator. This could be very cool.
when the release starts, the arpeggiator continues fading. This would be top for Surge.

**Comment by mkruselj** [comment]
Like all modulators, arpeggiator would indeed be per scene.

You can already do C64 like arps with the current Step Sequencer modulator from the LFO.

**Comment by Lo-lo78** [comment]
@mkruselj 
Yes, that's true but changing notes would be easier if they were included in the tracks envelopes
For instance:
switch between a major and a minor chord.
For the rest of the modulations it's not a problem.

**Comment by mkruselj** [comment]
You can already do this by copying the LFO to another LFO, changing the notes that you want, then crossfade the LFO amplitudes with a macro?

**Comment by Lo-lo78** [comment]
Thank you! I'll try to be more careful with the manual.
😂

**Comment by mkruselj** [comment]
A very good arpeggiator implementation can be found in certain Waldorf synths. Here's excerpt from Largo manual:

**Arp Mode:** _Normal / One Shot / Hold_

**Clock:** _1/64t…1280 bars_
Sets the note value for the steps of the rhythm pattern.

**Length:** _1/64t…1024 bars / legato_
Sets the length of the generated arpeggio notes. However, when Length is set to Legato, all arpeggio notes are played without pauses between each step and Arpeggiator Length therefore has no effect.

**Sort Order:** _as played / reversed / Key Lo>Hi / Key Hi>Lo / Vel Lo>Hi / Vel Hi>Lo_
Sets the order in which the note list is arranged. With this parameter, you can determine how the notes you originally play are split up for the arpeggio.

**Vel Mode:** _Each Note / First Note / Last Note_
Determines how velocity is interpreted in the arpeggio. Note that each arpeggio step might have an additional positive or negative offset set by the Arp Accent
parameter.
• If Each Note is selected, each note of the arpeggio is played back with the velocity that you originally played.
• If First Note is selected, the first note you played sets the velocity for all arpeggio steps.
• If Last Note is selected, the last note you played sets the velocity for all arpeggio steps.

**Direction:** _Up / Down / Alt Up / Alt Down_
Sets the direction that is used to play back the arpeggio.
• If Up is selected, the note list is played forward and the octaves are transposed upward. The arpeggio starts in the original octave and goes up to the highest octave. Then the arpeggio is repeated.
• If Down is selected, the note list is played backward and the octaves are transposed downward. The arpeggio starts in the highest octave and goes down to the original octave. Then the arpeggio is repeated.
• If Alt Up is selected, the note list is first played forward and the octaves are transposed upward. After reaching the last note of the note list in the
highest octave to play, the note list is played backward and the octaves are transposed downward down to the first note of the note list in the original octave. Then the arpeggio is repeated.
• If Alt Down is selected, the note list is first played backward and the octaves are transposed downward. The arpeggio starts in the highest octave. After reaching the first note of the note list in the original octave, the note list is played forward and the octaves are transposed upward up to the last note of the note list in the highest octave to play. Then the arpeggio is repeated.

**Octave Range:** _1...10_
Determines the range of the single notes in octaves. When it is set to 1 Oct, the note list will be played back in the same octave as originally entered Greater values mean that the note list is repeated in higher or lower octaves. The octave in which the arpeggio starts is determined by the Direction parameter. If you play notes that span more than one octave, they are still kept in the note list and played back before the note list is transposed.

**Max Notes:** _1...16_
Sets the maximum number of notes that are used for the note list. When you press more notes than set up here, the oldest notes are removed from the list. When you set it to 1 on a Bass sound, for example, the Arpeggiator only stores the last entered note and plays back sequentially. All other notes might still be used by other sounds, but the Bass sound plays back only this one particular note.

**Pattern Reset:** _off / on_
When all steps of an arpeggio pattern are played back, the pattern is repeated from the beginning so that the arpeggio is looped. With Pattern Reset, you can decide if the note list is also restarted from the beginning when the rhythm pattern is reset.


...and then there's the pattern editor, in which you can PER STEP adjust: velocity offset, timing offset, step duration, step direction, glide state

![image](https://github.com/surge-synthesizer/surge/assets/2393720/63ce9c57-d0af-4529-b16a-de1a2925350c)

**Step Direction** modes:
• If Normal (•) is selected, the Arpeggiator plays the step unaltered. The note list is advanced beforehand, except when you press a new chord.
• If Pause (o) is selected, the Arpeggiator plays nothing at this step position. When Length is set to legato, the previous step that isn’t set to pause is still held to create the legato effect. The note list is not advanced
• If Previous (<) is selected, the Arpeggiator plays the same note as it had to play in the previous step that was set to Normal or Random. With this setting, you can repeat a particular note of the note list several times. The note list is not advanced.
• If First (v) is selected, the Arpeggiator plays the very first note of the note list. This might be interesting if you want to only play the "root note" of a chord in a bass sound. The note list is not advanced.
• If Last (^) is selected, the Arpeggiator plays the very last note of the note list. The note list is not advanced.
• If First+Last (v^) is selected, the Arpeggiator plays a chord with two notes, the first and the last one of the note list. This means that you have to play at least two notes to hear the effect. Otherwise, you would hear only one note anyway. The note list is not advanced.
• If Chord is selected, the Arpeggiator plays a chord with all notes from the note list. This means that you have to play at least two notes to hear the effect. The note list is not advanced.
• If Random (?) random is selected, the Arpeggiator plays a random note from the note list. This doesn’t mean that it creates a note randomly; instead it uses one note of the existing note list. The note list is not advanced.
